SubresourceLoader should keep a ResourceResponse of the previous part of multipart/x-mixed-replace

fujii opened the pull request:
#### 3a699a9eda3a9a5744da1dd4ea8ef862ad56b1e2
<pre>
SubresourceLoader should keep a ResourceResponse of the previous part of multipart/x-mixed-replace
<a href="https://bugs.webkit.org/show_bug.cgi?id=263423">https://bugs.webkit.org/show_bug.cgi?id=263423</a>

Reviewed by Chris Dumez.

multipart/x-mixed-replace resources don&apos;t load progressively because
it keeps showing the previous part content while loading the next
part. Thus, SubresourceLoader finishes loading the current part
content when the next resource responce comes in.
SubresourceLoader::didReceiveResponse does the trick. It synthetically
calls CachedResource::finishLoading for the previous part content.

However, when CachedResource::finishLoading is called with the
previous part content, CachedResource::responseReceived is already
called with the next part resource responce. So, for example, if a
multipart/x-mixed-replace resource contains two parts, CachedResource
methods were called in the following sequence:

  CachedResource::responseReceived with multipart/x-mixed-replace
  CachedResource::responseReceived with the first part response
  CachedResource::responseReceived with the second part response
  CachedResource::finishLoading with the first part content
  CachedResource::finishLoading with the second part content

As the result, the last part doesn&apos;t show as expected.
&lt;<a href="https://webkit.org/b/36536">https://webkit.org/b/36536</a>&gt;

SubresourceLoader has to keep the previous ResourceResponse and
synthetically call CachedResource::responseReceived with it just
before synthetically calling CachedResource::finishLoading.

Also, this fixes another problem that debug build of Windows port was
failing an assertion for http/tests/multipart/invalid-image-data.html
test. The Image object wasn&apos;t recreated for the second part.
CachedImage::responseReceived clears the previous Image object. But,
it wasn&apos;t called for the second part just before finishLoading of the
second part.

* LayoutTests/platform/wincairo/TestExpectations:
* LayoutTests/platform/wincairo/http/tests/multipart/invalid-image-data-expected.txt:
* Source/WebCore/loader/SubresourceLoader.cpp:
* Source/WebCore/loader/SubresourceLoader.h:
* LayoutTests/http/tests/multipart/images-expected.html: Added.
* LayoutTests/http/tests/multipart/images.html: Added.
* LayoutTests/platform/glib/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/270023@main">https://commits.webkit.org/270023@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6c46904587675c9ce0b173eee6d1a2ab20e5d974

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24089 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2201 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25177 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26229 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22194 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24360 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3825 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24572 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22668 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24333 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1727 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20821 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26818 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1485 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21739 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27985 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21960 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22028 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25762 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1424 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/19095 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1445 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/21507 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5825 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1834 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1777 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -475,8 +475,25 @@ void SubresourceLoader::didReceiveResponse(const ResourceResponse& response, Com
-        m_resource->responseReceived(response);
+    if (m_loadingMultipartContent) {
+        if (!m_previousPartResponse.isNull()) {
+            if (m_resource) {
```
```
if (CachedResourceHandle resource = m_resource.get()) {
    resource->responseReceived(m_previousPartResponse);
    // The resource data will change as the next part is loaded, so we need to make a copy.
    resource->finishLoading(resourceData()->copy().ptr(), { });
}
```

Since you're modifying this code, you might as well convert it to our safer / modern smart pointer idioms.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -475,8 +475,25 @@ void SubresourceLoader::didReceiveResponse(const ResourceResponse& response, Com
+        m_documentLoader->subresourceLoaderFinishedLoadingOnePart(*this);
+        didFinishLoadingOnePart(emptyMetrics);
+    } else {
+        if (m_resource)
```
```
if (CachedResourceHandle resource = m_resource.get())
    resource->responseReceived(response);
```

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -475,8 +475,25 @@ void SubresourceLoader::didReceiveResponse(const ResourceResponse& response, Com
+        // Since a subresource loader does not load multipart sections progressively, data was delivered to the loader all at once.
+        // After the first multipart section is complete, signal to delegates that this load is "finished"
+        NetworkLoadMetrics emptyMetrics;
+        m_documentLoader->subresourceLoaderFinishedLoadingOnePart(*this);
```
Please add a new member function:
```
RefPtr<DocumentLoader> documentLoader() const { return m_documentLoader; }
```

Then use it here:
```
protectedDocumentLoader()->subresourceLoaderFinishedLoadingOnePart(*this);
```

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -750,6 +754,8 @@ void SubresourceLoader::didFinishLoading(const NetworkLoadMetrics& networkLoadMe
 
     m_state = Finishing;
+    if (m_loadingMultipartContent && !m_previousPartResponse.isNull())
+        m_resource->responseReceived(m_previousPartResponse);
```
Please add a new member function:
```
CachedResourceHandler<CachedResource> protectedCachedResource() const { return cachedResource(); }
```

then use it here:

```
protectedCachedResource()->responseReceived(m_previousPartResponse);
```

cdumez reviewed (approved):
LGTM but please adopt our modern smart pointer practices as suggested.

fujii committed:
SubresourceLoader should keep a ResourceResponse of the previous part of multipart/x-mixed-replace
https://bugs.webkit.org/show_bug.cgi?id=263423

Reviewed by Chris Dumez.

multipart/x-mixed-replace resources don't load progressively because
it keeps showing the previous part content while loading the next
part. Thus, SubresourceLoader finishes loading the current part
content when the next resource responce comes in.
SubresourceLoader::didReceiveResponse does the trick. It synthetically
calls CachedResource::finishLoading for the previous part content.

However, when CachedResource::finishLoading is called with the
previous part content, CachedResource::responseReceived is already
called with the next part resource responce. So, for example, if a
multipart/x-mixed-replace resource contains two parts, CachedResource
methods were called in the following sequence:

  CachedResource::responseReceived with multipart/x-mixed-replace
  CachedResource::responseReceived with the first part response
  CachedResource::responseReceived with the second part response
  CachedResource::finishLoading with the first part content
  CachedResource::finishLoading with the second part content

As the result, the last part doesn't show as expected.
<https://webkit.org/b/36536>

SubresourceLoader has to keep the previous ResourceResponse and
synthetically call CachedResource::responseReceived with it just
before synthetically calling CachedResource::finishLoading.

Also, this fixes another problem that debug build of Windows port was
failing an assertion for http/tests/multipart/invalid-image-data.html
test. The Image object wasn't recreated for the second part.
CachedImage::responseReceived clears the previous Image object. But,
it wasn't called for the second part just before finishLoading of the
second part.

* LayoutTests/platform/wincairo/TestExpectations:
* LayoutTests/platform/wincairo/http/tests/multipart/invalid-image-data-expected.txt:
* Source/WebCore/loader/SubresourceLoader.cpp:
* Source/WebCore/loader/SubresourceLoader.h:
* LayoutTests/http/tests/multipart/images-expected.html: Added.
* LayoutTests/http/tests/multipart/images.html: Added.
* LayoutTests/platform/glib/TestExpectations:

Canonical link: https://commits.webkit.org/270023@main
